### PR TITLE
Handler base class

### PR DESCRIPTION
**Pull request by kissgyorgy**
Big refactoring of the Handler implementation

### CODE REVIEW

File: unblob/models.py [hunk 1]
```diff
@@ -81,21 +85,40 @@ def __repr__(self):
...
+
+    def __init__(self):
+        if self.C_STRUCTURES:
+            self.cparser = cstruct()
+            self.cparser.load(self.C_STRUCTURES)
+
+            if self.HEADER_STRUCT:
+                self._header_parser = getattr(self.cparser, self.HEADER_STRUCT)
+
+    def parse_header(self, file: io.BufferedIOBase):
```
Comment (QKaiser) [inline unblob/models.py:110]: If we add an optional endianness argument that defaults to little endian, we could rely on parse_header in CPIO v3 handlers, and be ready for future use cases where we need to handle big endian transparently without rewriting parse_header in the subclass.

```suggestion
    def parse_header(self, file: io.BufferedIOBase, endian= "<"):
      self.cparser.endian = endian
```

File: unblob/handlers/filesystem/squashfs.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+        self.cparser_be.load(self.C_STRUCTURES)
+
+    def calculate_chunk(
+        self, file: io.BufferedIOBase, start_offset: int
+    ) -> Union[ValidChunk, UnknownChunk]:
+
+        # read the magic and derive endianness from it
+        magic_bytes = file.read(4)
+        magic = struct.unpack(">I", magic_bytes)[0]
+        cparser = self.cparser_be if magic == BIG_ENDIAN_MAGIC else self.cparser_le
```
Comment (QKaiser) [inline unblob/handlers/filesystem/squashfs.py:90]: If we use the optional endian parameter in parse_header in the Handler base class, we can do:

```suggestion
        endian = ">" if magic == BIG_ENDIAN_MAGIC else "<"
        header = self.parse_header(file, endian)
```

Of course we need to define `HEADER_STRUCT` for both CPIO v3.

**Review by QKaiser — CHANGES_REQUESTED** [review]
Just two comments related to endianness support in parse header, otherwise everything looks great !

File: unblob/models.py (hunk 1 above)
Comment (kissgyorgy) [reply unblob/models.py:110]: There will be only one `Handler` class instance during the whole run, so we still must not change any state of the instance. The __init__ is there, so the subclasses can customize some common behavior or initialize extra attributes needed.

Comment (vlaci) [reply unblob/models.py:110]: What is the overhead of `cparser` initialization? Is it worth storing it in the instance?

Comment (kissgyorgy) [reply unblob/models.py:110]:
```
>>> In [10]: %timeit parser=cstruct(); parser.load(C_STRUCTURES)
2.17 ms ± 63.1 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Review by kukovecz — APPROVED** [review]
The refactoring looks good to me. From my PoV, this can be merged (when concerns raised by others are attended)

File: unblob/models.py (hunk 1 above)
Comment (QKaiser) [reply unblob/models.py:110]: Ok that makes sense. Let's leave it this way then.

**Review by QKaiser — APPROVED** [review]

**Pull request merged**